src/hooks/op_rewrite.c

```diff
@@ -0,0 +1,262 @@
+#include <postgres.h>
+
+#include "op_rewrite.h"
+
+#include <access/genam.h>
+#include <assert.h>
+#include <catalog/pg_amproc.h>
+#include <catalog/pg_opclass.h>
+#include <commands/defrem.h>
+#include <miscadmin.h>
+#include <nodes/nodeFuncs.h>
+#include <parser/parsetree.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <utils/rel.h>
+#include <utils/syscache.h>
+
+#include "plan_tree_walker.h"
+#include "utils.h"
+
+#if PG_VERSION_NUM < 120000
+#include <access/heapam.h>
+#include <access/htup_details.h>
+#else
+#include <access/relation.h>
+#endif
+
+static Node *operator_rewriting_mutator(Node *node, void *ctx);
+
+void base_plan_mutator(Plan *plan, void *context)
+{
+    plan->targetlist = (List *)operator_rewriting_mutator((Node *)plan->targetlist, context);
+    plan->qual = (List *)operator_rewriting_mutator((Node *)plan->qual, context);
+    plan->lefttree = (Plan *)operator_rewriting_mutator((Node *)plan->lefttree, context);
+    plan->righttree = (Plan *)operator_rewriting_mutator((Node *)plan->righttree, context);
+    plan->initPlan = (List *)operator_rewriting_mutator((Node *)plan->initPlan, context);
+}
+
+Node *plan_tree_mutator(Plan *plan, void *context)
+{
+    check_stack_depth();
+
+    switch(nodeTag(plan)) {
```

Comment:
Question: Are these all of the cases that need to be covered?

Reply:
I'm not familiar enough with all of them to say for sure. I removed some I know are definitely unnecessary. The remaining ones I couldn't definitively rule out so I was hoping to get some additional eyes on them 

some very rough and possibly erroneous notes.  

- `SubqueryScan`  I think this is required, since a subquery could include the case we're trying to address
- `CteScan` Unecessary if the CTE is the created table it's unnecessary. Necessary if it's the logic to create the table. I believe it's the former but am not certain
- `Join` One of both of the tables being joined can be generated by a select so I think this is required
- `Agg` I think that this can cover what is essentially a subquery e.g. `ANY(SELECT ....)`
- `Group` This can probably be removed, I wasn't sure if it could have a subquery with an order by but at least for a single query I think it's 1 or the other 
- `Sort` I think `ORDDER BY` can generate a sort meriting its inclusion, but I'm not sure 
- `Unique`This can probably be removed, I don't think you can call unique on an expression, unless you can do `unique(v <-> ...)` 
- `NestLoop` loops can be seqscans, I think this is required
- `Result` I couldn't find much information on this plan type, and so included it to be safe
- `Limit` When I looked at this it seemed like the limit becomes the root node, so I think we need this
- `Append`Not 100% sure where this will end up in the tree, but I think you can have a union of two scans so this should be checked

Reply:
Haha just realized I misunderstood the question, apologies. I had assumed that plan tree walker was covering and just removed the nodes I have explicit checks for. I don't know enough about the remaining plan types to say if they might require this. I will try to read more about them. Immediately `T_IncrementalSort` feels like it should probably be included so my guess is that I'm missing several others

Reply:
No worries, I was just asking if this was a deliberately curated subset or just what you ended up covering. I don't know if removing is unnecessary, since I know the operator can end up in funny subqueries and joins.

Assuming you meant to refer to the plan walker I wrote -- there's around 30 plan nodes I believe, which you can see in nodes.h. I only covered the ones that I was able to generate tests for. It probably would be better for me to have added the other cases as well. I just couldn't figure out tests for the rest. For example, I couldn't trigger T_Join (joins ended up using a different plan node), and it ended up getting removed in Postgres 16. I'm not sure if it gets triggered at all in Postgres 15.

Reply:
Ah, that makes sense. You got way more than I've been able to generate! Reading through the additional types it seems like most  if not all of the ones that were left out can be safely ignored. I'll try going over the rest of them tomorrow. It's funny that T_Join is hard to trigger, you'd expect it would be relatively common. 

One option I could implement pretty easily is to call the base mutator in the default so that we get some coverage on the more exotic types, this could be combined with a check that returns the path as is for ones we think are safe to ignore too.